src/Workspaces/Core/Portable/CodeFixes/CodeFix.cs

```diff
@@ -38,17 +42,41 @@ internal sealed class CodeFix
 
         internal CodeFix(Project project, CodeAction action, Diagnostic diagnostic)
         {
-            this.Project = project;
-            this.Action = action;
-            this.Diagnostics = ImmutableArray.Create(diagnostic);
+            Project = project;
+            Action = action;
+            Diagnostics = ImmutableArray.Create(diagnostic);
         }
 
         internal CodeFix(Project project, CodeAction action, ImmutableArray<Diagnostic> diagnostics)
         {
             Debug.Assert(!diagnostics.IsDefault);
-            this.Project = project;
-            this.Action = action;
-            this.Diagnostics = diagnostics;
+            Project = project;
+            Action = action;
+            Diagnostics = diagnostics;
+        }
+
+        internal DiagnosticData GetPrimaryDiagnosticData()
+        {
+            var diagnostic = PrimaryDiagnostic;
+
+            if (diagnostic.Location.IsInSource)
+            {
+                var document = Project.GetDocument(diagnostic.Location.SourceTree);
+                if (document != null)
+                {
+                    return DiagnosticData.Create(diagnostic, document);
+                }
+            }
+            else if (diagnostic.Location.Kind == LocationKind.ExternalFile)
+            {
+                var document = Project.Documents.FirstOrDefault(d => d.FilePath == diagnostic.Location.GetLineSpan().Path);
```

Comment:
`Project.Documents` does not include `AdditionalDocuments` or `AnalyzerConfigDocuments`. Should we check them as well?